File: sdk/java/src/main/java/supply/escs/ESCSTest.java
package supply.escs;

import java.time.Instant;
import java.util.List;

/**
 * ESCS Java SDK — live integration tests
 * Run: mvn test  or  javac + java manually
 */
public class ESCSTest {

    static int passed = 0;
    static int failed = 0;

    static void test(String name, boolean condition) {
        if (condition) {
            System.out.println("  PASS: " + name);
            passed++;
        } else {
            System.out.println("  FAIL: " + name);
            failed++;
        }
    }

    public static void main(String[] args) throws Exception {
        ESCSClient client = new ESCSClient("http://127.0.0.1:7710");
        long now = Instant.now().getEpochSecond();

        System.out.println("=== ESCS Java SDK Tests ===\n");

        // Health
        ESCSClient.Health h = client.health();
        test("health ok", h.isOk());
        test("health service is adapterd", "adapterd".equals(h.getService()));
        test("health node_id present", h.getNodeId() != null && h.getNodeId().startsWith("ed25519:"));

        // Batch created
        ESCSClient.Receipt r = client.batchCreated(
            ESCSClient.BatchCreatedRequest.builder()
                .batchId("batch:JAVA-TEST-001")
                .productCode("DRUG-001")
                .quantity(5000)
                .unit("units")
                .originLocation("Chicago, IL")
                .producerId("party:producer")
                .lotNumber("LOT-JAVA-001")
                .manufactureDate(now)
                .expiryDate(now + 31536000L)
                .build()
        );
        test("batch_created ok", r.isOk());
        test("batch_created published", r.isPublished());
        test("batch_created witnessed", r.isWitnessed());
        test("batch_created digest starts sha256:", r.getDigest() != null && r.getDigest().startsWith("sha256:"));
        test("batch_created claim_space is SUPPLY.BATCH.v1", "SUPPLY.BATCH.v1".equals(r.getClaimSpace()));

        // Custody transfer
        r = client.custodyTransfer(
            ESCSClient.CustodyTransferRequest.builder()
                .batchId("batch:JAVA-TEST-001")
                .from("party:producer")
                .to("party:shipper")
                .location("Chicago, IL")
                .quantity(5000)
                .unit("units")
                .handoffMethod("truck")
                .build()
        );
        test("custody_transfer ok", r.isOk());
        test("custody_transfer published", r.isPublished());
        test("custody_transfer claim_space", "SUPPLY.CUSTODY.v1".equals(r.getClaimSpace()));

        // Customs cleared — rep=10 passes
        r = client.customsCleared(
            ESCSClient.CustomsClearedRequest.builder()
                .batchId("batch:JAVA-TEST-001")
                .customsAuthority("CBP")
                .clearanceRef("CBP-JAVA-001")
                .portOfEntry("JFK")
                .destinationCountry("US")
                .reputation(10)
                .build()
        );
        test("customs_cleared ok", r.isOk());
        test("customs_cleared published", r.isPublished());

        // Customs cleared — rep=9 denied
        r = client.customsCleared(
            ESCSClient.CustomsClearedRequest.builder()
                .batchId("batch:JAVA-TEST-001")
                .customsAuthority("CBP")
                .clearanceRef("CBP-JAVA-002")
                .portOfEntry("JFK")
                .destinationCountry("US")
                .reputation(9)
                .build()
        );
        test("customs gate denied at rep=9", !r.isPublished());

        // Temperature log
        r = client.temperatureLog(
            ESCSClient.TemperatureLogRequest.builder()
                .batchId("batch:JAVA-TEST-001")
                .sensorId("sensor:cold-007")
                .location("Chicago Cold Storage")
                .tempMinC(2.1)
                .tempMaxC(7.8)
                .tempAvgC(4.5)
                .windowStart(now - 299)
                .windowEnd(now)
                .readingCount(60)
                .reputation(1)
                .timestamp(now)
                .build()
        );
        test("temperature_log ok", r.isOk());
        test("temperature_log published", r.isPublished());

        // Recall issued — rep=50 passes
        r = client.recallIssued(
            ESCSClient.RecallIssuedRequest.builder()
                .recallId("recall:JAVA-TEST-001")
                .affectedBatchIds(List.of("batch:JAVA-TEST-001"))
                .affectedLotNumbers(List.of("LOT-JAVA-001"))
                .productCode("DRUG-001")
                .severity("class_i")
                .reason("java sdk test")
                .issuerId("oracle:fda-001")
                .instructions("destroy")
                .regulatoryRef("FDA-JAVA-2025-001")
                .reputation(50)
                .build()
        );
        test("recall_issued ok", r.isOk());
        test("recall_issued published", r.isPublished());
        test("recall_issued witnessed", r.isWitnessed());

        // Recall gate denied — rep=49
        r = client.recallIssued(
            ESCSClient.RecallIssuedRequest.builder()
                .recallId("recall:JAVA-TEST-002")
                .affectedBatchIds(List.of("batch:JAVA-TEST-001"))
                .affectedLotNumbers(List.of("LOT-JAVA-001"))
                .productCode("DRUG-001")
                .severity("class_i")
                .reason("should be denied")
                .issuerId("oracle:fda-001")
                .instructions("destroy")
                .regulatoryRef("FDA-JAVA-2025-002")
                .reputation(49)
                .build()
        );
        test("recall gate denied at rep=49", !r.isPublished());

        // Receipt URL present
        r = client.custodyTransfer(
            ESCSClient.CustodyTransferRequest.builder()
                .batchId("batch:JAVA-TEST-001")
                .from("party:shipper")
                .to("party:distributor")
                .location("New York, NY")
                .quantity(5000)
                .unit("units")
                .build()
        );
        test("receipt_url present", r.getReceiptUrl() != null && r.getReceiptUrl().startsWith("http"));

        System.out.println("\n=========================");
        System.out.println("  total  " + passed + " passed  " + failed + " failed");
        System.out.println("=========================");
        if (failed > 0) System.exit(1);
    }
}

File: sdk/go/escs_test.go
package escs_test

import (
"context"
"testing"
"time"

escs "github.com/mauludsadiq/ESCS/sdk/go"
)

var client = escs.NewClient("http://localhost:7710")
var ctx = context.Background()
var now = time.Now().Unix()

func TestHealth(t *testing.T) {
h, err := client.Health(ctx)
if err != nil {
t.Fatalf("health check failed: %v", err)
}
if !h.IsOk() {
t.Fatal("health not ok")
}
if h.GetService() == "" {
t.Fatal("missing service name")
}
t.Logf("node: %s", h.GetNodeID()[:20])
}

func TestJurisdictions(t *testing.T) {
j, err := client.Jurisdictions(ctx)
if err != nil {
t.Fatalf("jurisdictions failed: %v", err)
}
if j.Count != 42 {
t.Fatalf("expected 42 jurisdictions, got %d", j.Count)
}
}

func TestEventTypes(t *testing.T) {
e, err := client.EventTypes(ctx)
if err != nil {
t.Fatalf("event types failed: %v", err)
}
if e.Count != 18 {
t.Fatalf("expected 18 event types, got %d", e.Count)
}
}

func TestBatchCreated(t *testing.T) {
r, err := client.BatchCreated(ctx, escs.BatchCreatedRequest{
BatchID:         "batch:GO-TEST-001",
ProductCode:     "DRUG-001",
Quantity:        1000,
Unit:            "units",
OriginLocation:  "Chicago, IL",
ProducerID:      "party:producer",
LotNumber:       "LOT-GO-001",
ManufactureDate: now,
ExpiryDate:      now + 31536000,
})
if err != nil {
t.Fatalf("batch_created failed: %v", err)
}
if !r.OK {
t.Fatalf("not ok: %s", r.Error)
}
if !r.Published {
t.Fatal("not published")
}
if !r.Witnessed {
t.Fatal("not witnessed")
}
if r.Digest == "" {
t.Fatal("missing digest")
}
t.Logf("digest: %s", r.Digest[:20])
}

func TestCustodyTransfer(t *testing.T) {
r, err := client.CustodyTransfer(ctx, escs.CustodyTransferRequest{
BatchID:  "batch:GO-TEST-001",
From:     "party:producer",
To:       "party:shipper",
Location: "Chicago, IL",
Quantity: 1000,
Unit:     "units",
})
if err != nil {
t.Fatalf("custody_transfer failed: %v", err)
}
if !r.OK || !r.Published {
t.Fatalf("not published: %s", r.Error)
}
if r.ClaimSpace != "SUPPLY.CUSTODY.v1" {
t.Fatalf("wrong claim_space: %s", r.ClaimSpace)
}
}

func TestInspectionPassed(t *testing.T) {
r, err := client.InspectionPassed(ctx, escs.InspectionPassedRequest{
BatchID:        "batch:GO-TEST-001",
InspectorID:    "inspector:fda-001",
InspectionType: "fda_gmp",
Location:       "Chicago, IL",
Standards:      []string{"FDA-21CFR"},
Notes:          "all clear",
Reputation:     5,
})
if err != nil {
t.Fatalf("inspection_passed failed: %v", err)
}
if !r.OK || !r.Published {
t.Fatalf("not published: %s", r.Error)
}
}

func TestInspectionGateDenied(t *testing.T) {
r, err := client.InspectionPassed(ctx, escs.InspectionPassedRequest{
BatchID:        "batch:GO-TEST-001",
InspectorID:    "inspector:fda-001",
InspectionType: "fda_gmp",
Location:       "Chicago, IL",
Reputation:     4, // below RepMin(5)
})
if err != nil {
t.Fatalf("request failed: %v", err)
}
if r.Published {
t.Fatal("should have been gate denied at rep=4")
}
}

func TestCustomsCleared(t *testing.T) {
r, err := client.CustomsCleared(ctx, escs.CustomsClearedRequest{
BatchID:            "batch:GO-TEST-001",
CustomsAuthority:   "CBP",
ClearanceRef:       "CBP-GO-TEST-001",
PortOfEntry:        "JFK",
DestinationCountry: "US",
Reputation:         10,
})
if err != nil {
t.Fatalf("customs_cleared failed: %v", err)
}
if !r.OK || !r.Published {
t.Fatalf("not published: %s", r.Error)
}
}

func TestTemperatureLog(t *testing.T) {
tsNow := time.Now().Unix()
r, err := client.TemperatureLog(ctx, escs.TemperatureLogRequest{
BatchID:      "batch:GO-TEST-001",
SensorID:     "sensor:cold-007",
Location:     "Chicago Cold Storage",
TempMinC:     2.1,
TempMaxC:     7.8,
TempAvgC:     4.5,
WindowStart:  tsNow - 299,
WindowEnd:    tsNow,
ReadingCount: 60,
Reputation:   1,
Timestamp:    tsNow,
})
if err != nil {
t.Fatalf("temperature_log failed: %v", err)
}
if !r.OK || !r.Published {
t.Fatalf("not published: %s", r.Error)
}
}

func TestTemperatureBreach(t *testing.T) {
r, err := client.TemperatureBreach(ctx, escs.TemperatureBreachRequest{
BatchID:         "batch:GO-TEST-001",
SensorID:        "sensor:cold-007",
Location:        "Chicago Cold Storage",
TempC:           11.2,
ThresholdMinC:   2.0,
ThresholdMaxC:   8.0,
BreachTimestamp: now - 600,
DurationSecs:    600,
Reputation:      1,
})
if err != nil {
t.Fatalf("temperature_breach failed: %v", err)
}
if !r.OK || !r.Published {
t.Fatalf("not published: %s", r.Error)
}
}

func TestCertificationIssued(t *testing.T) {
r, err := client.CertificationIssued(ctx, escs.CertificationIssuedRequest{
BatchID:     "batch:GO-TEST-001",
CertType:    "fda_gmp",
CertifierID: "certifier:fda-001",
CertRef:     "FDA-GMP-GO-001",
ValidFrom:   now,
ValidUntil:  now + 31536000,
Reputation:  20,
})
if err != nil {
t.Fatalf("certification_issued failed: %v", err)
}
if !r.OK || !r.Published {
t.Fatalf("not published: %s", r.Error)
}
}

func TestRecallIssued(t *testing.T) {
r, err := client.RecallIssued(ctx, escs.RecallIssuedRequest{
RecallID:           "recall:GO-TEST-001",
AffectedBatchIDs:   []string{"batch:GO-TEST-001"},
AffectedLotNumbers: []string{"LOT-GO-001"},
ProductCode:        "DRUG-001",
Severity:           "class_i",
Reason:             "go sdk test recall",
IssuerID:           "oracle:fda-001",
Instructions:       "destroy",
RegulatoryRef:      "FDA-GO-2025-001",
Reputation:         50,
})
if err != nil {
t.Fatalf("recall_issued failed: %v", err)
}
if !r.OK || !r.Published {
t.Fatalf("not published: %s", r.Error)
}
}

func TestRecallGateDenied(t *testing.T) {
r, err := client.RecallIssued(ctx, escs.RecallIssuedRequest{
RecallID:           "recall:GO-TEST-002",
AffectedBatchIDs:   []string{"batch:GO-TEST-001"},
AffectedLotNumbers: []string{"LOT-GO-001"},
ProductCode:        "DRUG-001",
Severity:           "class_i",
Reason:             "should be denied",
IssuerID:           "oracle:fda-001",
Instructions:       "destroy",
RegulatoryRef:      "FDA-GO-2025-002",
Reputation:         49, // below RepMin(50)
})
if err != nil {
t.Fatalf("request failed: %v", err)
}
if r.Published {
t.Fatal("should have been gate denied at rep=49")
}
}

func TestOriginAttested(t *testing.T) {
r, err := client.OriginAttested(ctx, escs.OriginAttestedRequest{
BatchID:         "batch:GO-TEST-001",
CertifierID:     "certifier:origin-001",
CountryOfOrigin: "US",
Region:          "Illinois",
Standards:       []string{"FDA-21CFR"},
Reputation:      10,
})
if err != nil {
t.Fatalf("origin_attested failed: %v", err)
}
if !r.OK || !r.Published {
t.Fatalf("not published: %s", r.Error)
}
}

func TestProvenance(t *testing.T) {
p, err := client.Provenance(ctx, "batch:GO-TEST-001")
if err != nil {
t.Fatalf("provenance failed: %v", err)
}
if !p.OK {
t.Fatal("provenance not ok")
}
if p.BatchID != "batch:GO-TEST-001" {
t.Fatalf("wrong batch_id: %s", p.BatchID)
}
}
